### PR TITLE
feat: Query Parquet files

### LINKED ISSUE #1082: Query data from Object Store
__Use case:__
When data is persisted in object store that is done in #808, our Query Server needs to read that data from object store if they no longer exist in Read Buffer. We will see which of the following we will implement first

 1. Will we query parquet files directly or first load them into the read buffer
 2. Do we need to persist a serialized catalog for v1 or can we just list objectstore files

TBD: How to do it 

__Proposal:__
Short summary of the feature.

__Current behaviour:__
Describe what currently happens.

__Desired behaviour:__
Describe what you want.

__Alternatives considered:__
Describe other solutions or features you considered.

### PR DESCRIPTION
**Pull request by NGA-TRAN**
Closes #1082

Describe your proposed changes here.

- [ ] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [ ] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: parquet_file/src/utils.rs [hunk 1]
```diff
@@ -436,13 +486,48 @@ fn make_record_batch() -> (Vec<RecordBatch>, Schema, Vec<ColumnSummary>, Timesta
...
+    let mut arrow_reader = ParquetFileArrowReader::new(Arc::new(reader));
+
+    // Indices of columns in the schema needed to read
+    let projection: Vec<usize> = Storage::column_indices(Selection::All, Arc::clone(&schema));
+    let mut batch_reader = arrow_reader
+        .get_record_reader_by_columns(projection, 1024)
+        .unwrap();
+    loop {
+        match batch_reader.next() {
+            Some(Ok(batch)) => {
```
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply parquet_file/src/utils.rs:515]: This function is written for testing only

File: parquet_file/src/storage.rs [hunk 2]
```diff
@@ -305,10 +330,15 @@ impl Storage {
...
+        // Schema of all record batches must be the same, Get the first one
+        // to build record batch stream
+        let batch_schema = if batches.is_empty() {
+            schema
+        } else {
+            batches[0].schema()
+        };
 
-        Ok(Box::pin(SizedRecordBatchStream::new(schema, batches)))
+        Ok(Box::pin(SizedRecordBatchStream::new(batch_schema, batches)))
```
Comment (NGA-TRAN) [inline parquet_file/src/storage.rs:341]: This is the second bug fix and as a result, all the tests passed

File: server/src/query_tests/scenarios.rs [hunk 3]
```diff
@@ -374,20 +374,19 @@ impl DbSetup for EndToEndTest {
     }
 }
 
-/// This function loads two chunks of lp data into 4 different scenarios
+/// This function loads one chunk of lp data into different scenarios that simulates
+/// the data life cycle.
```
Comment (NGA-TRAN) [inline server/src/query_tests/scenarios.rs:378]: By introducing scenario5 in one_chunk case and scenario6 in 2-chunk case, we have added a lot of e2e tests for reading data in object store only.

Note that, we currently only support files in local disk. Will work on supporting all kind of storage next. Ticket #1342

**Comment by alamb** [comment]
checking it out

File: server/src/query_tests/scenarios.rs [hunk 4]
```diff
@@ -433,10 +434,29 @@ pub(crate) async fn make_one_chunk_scenarios(partition_key: &str, data: &str) ->
...
+        db.rollover_partition(partition_key, &table_name)
+            .await
+            .unwrap();
+        db.load_chunk_to_read_buffer(partition_key, &table_name, 0)
+            .await
+            .unwrap();
+        db.write_chunk_to_object_store(partition_key, &table_name, 0)
+            .await
+            .unwrap();
+        db.unload_read_buffer(partition_key, &table_name, 0)
```
Comment (alamb) [inline server/src/query_tests/scenarios.rs:450]: 😎  Nice

File: server/src/query_tests/scenarios.rs (hunk 3 above)
Comment (alamb) [reply server/src/query_tests/scenarios.rs:378]: I am quite pleased with how the `query_test` harness is working out.

**Review by alamb — APPROVED** [review]
I think it would help to remove much of the commented out code  `//println!`  but the code that is here I think is pretty much good to go.

I think there is one `unwrap()` that should probably be made into a real error but that could  be handled in a subsequent PR as well I think

This code and its comments make a lot of sense to me. Nice work @NGA-TRAN !

File: parquet_file/src/storage.rs (hunk 2 above)
Comment (NGA-TRAN) [reply parquet_file/src/storage.rs:341]: @alamb : Yes. The schema got changed (I did not look the exact different but I think the index of the column) and also, it makes sense to use the schema that gos with the batch

File: parquet_file/src/storage.rs [hunk 5]
```diff
@@ -456,6 +507,20 @@ impl Storage {
...
     }
+
+    pub async fn load_parquet_data_from_object_store(
+        path: Path,
+        store: Arc<ObjectStore>,
+    ) -> Result<Vec<u8>> {
+        store
+            .get(&path)
+            .await
+            .unwrap()
```
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply parquet_file/src/storage.rs:518]: Yeah. I missed this :) This was moved from a test-only file that is why unwrap was used

File: parquet_file/src/storage_testing.rs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        let parquet_metadata = read_parquet_metadata_from_file(parquet_data).unwrap();
+        //
+        // 1. Check metadata at file level: Everything is correct
+        let schema_actual = read_schema_from_parquet_metadata(&parquet_metadata).unwrap();
+        //let schema_expected = chunk.table_schema(&table, Selection::All).unwrap();
+
+        // println!("---- Actual file level key value metadata: {:#?}", schema_actual.as_arrow().metadata().clone());
+        // println!("---- Actual schema: {:#?}", schema_actual);
+        assert_eq!(schema.clone(), schema_actual);
+        assert_eq!(
```
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply parquet_file/src/storage_testing.rs:60]: Right. The line above does include metadata

File: server/src/db.rs [hunk 7]
```diff
@@ -785,6 +790,12 @@ impl Db {
                 streams::ReadFilterResultsStream::new(read_results, Arc::clone(&arrow_schema)),
             );
 
+            // let mut batches = Vec::new();
```
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply server/src/db.rs:793]: Hah, I missed removing them

File: parquet_file/src/utils.rs [hunk 8]
```diff
@@ -436,13 +486,48 @@ fn make_record_batch() -> (Vec<RecordBatch>, Schema, Vec<ColumnSummary>, Timesta
...
+
+pub fn read_data_from_parquet_data(schema: SchemaRef, parquet_data: Vec<u8>) -> Vec<RecordBatch> {
+    let mut record_batches = vec![];
+
+    let cursor = SliceableCursor::new(parquet_data);
+    let reader = SerializedFileReader::new(cursor).unwrap();
+    let mut arrow_reader = ParquetFileArrowReader::new(Arc::new(reader));
+
+    // Indices of columns in the schema needed to read
+    let projection: Vec<usize> = Storage::column_indices(Selection::All, Arc::clone(&schema));
```
Comment (carols10cents) [inline parquet_file/src/utils.rs:509]: This type annotation isn't necessary either, if you like having it here, that's fine though :)
```suggestion
    let projection = Storage::column_indices(Selection::All, Arc::clone(&schema));
```

File: server/src/db/chunk.rs [hunk 9]
```diff
@@ -14,6 +14,9 @@ use observability_deps::tracing::debug;
 use parquet_file::chunk::Chunk as ParquetChunk;
 use query::{exec::stringset::StringSet, predicate::Predicate, PartitionChunk};
 use read_buffer::Chunk as ReadBufferChunk;
+//use read_buffer::ReadFilterResults;
+
+//use futures::StreamExt;
```
Comment (carols10cents) [inline server/src/db/chunk.rs:19]: Some more commented-out code here

**Review by carols10cents — COMMENTED** [review]
This looks like it was tricky to figure out and test thoroughly! I just have a few cleanup suggestions :)

File: parquet_file/src/storage.rs [hunk 10]
```diff
@@ -1,37 +1,45 @@
...
 use std::{
+    convert::TryInto,
     fs::File,
     io::{Cursor, Seek, SeekFrom, Write},
     sync::Arc,
     task::{Context, Poll},
-};
+}; // ops::Deref,
+   // use tokio::sync::mpsc::{channel, Receiver, Sender};
+   // use tokio::task;
```
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply parquet_file/src/storage.rs]: They should be removed

File: parquet_file/src/storage.rs [hunk 11]
```diff
@@ -79,10 +87,26 @@ pub enum Error {
...
         source: datafusion::error::DataFusionError,
     },
+
+    #[snafu(display("Cannot read arrow schema from parquet: {}", source))]
+    ArrowFromParquetFailure {
+        source: parquet::errors::ParquetError,
+    },
+
+    #[snafu(display("Cannot read IOx schema from arrow: {}", source))]
+    IoxFromArrowFailure {
```
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply parquet_file/src/storage.rs:102]: Great catch

File: parquet_file/src/storage_testing.rs [hunk 12]
```diff
@@ -0,0 +1,109 @@
+/// This is test module of the storage module
+// The tests have their own module to avoid cycle "use"
```
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply parquet_file/src/storage_testing.rs:2]: Yes, Carol. Please push. I have just pushed some changes, too

**Pull request merged**